### ISSUE TITLE
`eth_gasPrice` should not return a value smaller than `parity_minGasPrice`.

### ISSUE THREAD

**Issue by MicahZoltu** [paritytech/parity#5396]
If the Parity node has a `parity_minGasPrice` set, it should not return an `eth_gasPrice` that is lower than it will accept.

At the moment, if the average gas price for the last 100 transactions is 0 and you call `parity_setMinGasPrice` with `0x2` parameter, `eth_gasPrice` will return `0x0`.  Any dApp that uses this to determine how much gas to supply transactions with (with I believe is the common use case for it) will fail to submit the transaction even to the local node because the gas price is too low.

This is particularly a pain in dev/private chain environment where gas price average over the last 100 blocks may be 0 but the min gas price may be substantially higher than that (e.g., 20GWei).

Side Note: If someone knows how I can force up the value that `eth_gasPrice` returns please let me know.  I tried submitting a transaction with a very high gas price but that doesn't appear to have any impact on `eth_gasPrice`'s return value.  Perhaps it is filtering out outliers?

**Comment by tomusdrw** [comment]
Local transactions are always accepted to the queue even if it's below minimal gas price. IMHO `minGasPrice` is a bit disjoint setting, it's a threshold required by external transactions (those propagated over the network) and it's only useful for miners. Your local transactions can use any `gasPrice` you want. `eth_gasPrice` is estimating the sensible gas price for your (local) transaction based on the network median (that's why submitting single transaction with high gas price won't work if you have other transactions that were mined with lower gas prices).

So in summary:
1. For your specific use case submitting transactions with low gas price (even 0) should work really well (local node = mining node). If it's not we need to do a bit more digging to understand what's going wrong.
2. If you think there is a minimal sensible gas price that should be used you can always do `Math.max(web3.eth.gasPrice, myMinGasPrice)`
3. I think Signer UI might be a good place to inform user the gas price they are trying to set might be too low, that's a separate issue (@jacogr what do you think about that?).

**Comment by MicahZoltu** [comment]
Thanks for giving the clarity on the difference.  There may be value in improving the Wiki documentation on these calls to make it more clear to the reader what exactly they do (`parity_minGasPrice`) and/or where the data is sourced from (`eth_gasPrice`).  Before posting this issue, I did a lot of manual troubleshooting and read the docs several times but I still came to the wrong conclusion.

I also learned (via Parity Gitter) that Parity _currently_ calculates `eth_gasPrice` from the last 100 transactions, so if you want to force the `eth_gasPrice` to something (e.g., for testing) you need to send 51 transactions with the desired price for every 49 normal transactions.  It would be nice if there were an easy way to override the whole median of last 100 transactions via RPC or CLI, but at least now that I understand the situation I can script it in my test infrastructure/dev setup.

Re: My situation: it is complicated by the fact that the dApp tries to smartly set the gas price appropriately and the contracts expect extra gas via value to fund future calls and they throw on `0` (basically `0` is magic at the moment).  I resolved my problem by calling a single `eth_sendTransaction` from/to the only account with 0 value and `gasPrice 1` before doing anything else.  Since the rest of the code relies on `eth_gasPrice` to figure out the gas price they use this results in everything using `gasPrice 1`.  Really it was all just about understanding what these things were doing under the hood and I was able to resolve my problem pretty easily, hence my suggestion above about improving documentation.  :)

**Comment by tomusdrw** [comment]
Actually Parity calculates `eth_gasPrice` as a median of transactions in last 100 blocks. So pretty much you need to send one transaction with gas price higher than 0. If you already have 100 transactions with gas price 0, then yes, you would need to send 51 transactions to rise the value returned by `eth_gasPrice`.

The idea of `eth_gasPrice` is to estimate sensible gas price in the network taking into account other transactions already included in blocks. There is a `parity_gasPriceHistogram` method returning more detailed info about gas price of transactions in last 100 blocks. You can also customize gas price in the UI.

I think in your setup the problem is that even the first transaction you send was using `eth_gasPrice` and it was just set to `0` causing all other transactions to get the same gas price in future. @keorn maybe we should return `minGasPrice` instead of `0` if there are no transactions found in last 100 blocks?

**Comment by 5chdn** [comment]
Bumping this discussion to make sure we move on.

@keorn should we return `minGasPrice` instead of `0` if there are no transactions found in last 100 blocks? Please try to add a comment within 7 days. Also, if you are consent on a solution, please assign this ticket and decide on a deadline.

**Comment by sjeohp** [comment]
Any update on this?  I don't mind making the change if it is required.

**Comment by keorn** [comment]
`minGasPrice` is only for mining, but indeed it makes out of the box setup annoying. Rather than returning it for user, I would set up a different `minGasPrice` default for dev/poa chains using https://github.com/paritytech/parity/issues/4673

**Comment by sjeohp** [comment]
`parity preset dev` in https://github.com/paritytech/parity/pull/6044 has a 0 `min_gas_price`
